In [1]:
import json
import torch
import torch.nn as nn
import random
from tqdm import tqdm
from transformers import GPT2LMHeadModel, GPT2Tokenizer

In [2]:
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')

In [3]:
dev_dict = json.load(open('../data/coqa-dev-v1.0.json', encoding='utf8'))

In [4]:
train_list = json.load(open('../data/qa_train_list.json', encoding='utf8'))
dev_list = json.load(open('../data/qa_dev_list.json', encoding='utf8'))

In [5]:
def chunks(lst, n):
    for i in range(0, len(lst), n):
        yield lst[i:i + n]

def batchify(data, n):
    len_dict = {}
    for item in data:
        length = item.shape[1]
        try:
            len_dict[length].append(item)
        except:
            len_dict[length] = [item]

    batch_chunks = []
    for k in len_dict.keys():
        vectors = len_dict[k]
        batch_chunks += chunks(vectors, n)

    batches = []
    for chunk in batch_chunks:
        inputs = torch.stack([item[0] for item in chunk])
        batches.append((inputs))

    return batches

In [6]:
_limit = 1024
data = []
total_trimmed = 0
for item in train_list:
    tokens = tokenizer.encode(item, return_tensors='pt')
    if tokens.shape[1] > _limit:
        tokens = tokens[:, :_limit]
        total_trimmed += 1
    data.append(tokens)
print(f'Trimmed {total_trimmed} out of {len(train_list)}')

Token indices sequence length is longer than the specified maximum sequence length for this model (1088 > 1024). Running this sequence through the model will result in indexing errors


Trimmed 52 out of 7199


In [6]:
train_batches = batchify(data, 1)

NameError: name 'data' is not defined

In [5]:
model = GPT2LMHeadModel.from_pretrained('gpt2')
model.cuda()
criterion = nn.BCEWithLogitsLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=2e-5)

In [8]:
def train(train_model, batches, optimizer, criterion):
    total_loss = 0.
    for i, batch in tqdm(enumerate(batches), total=len(batches)):
        model.train()
        inputs = batch
        optimizer.zero_grad()
        loss = train_model(inputs.cuda(), labels=inputs.cuda())[0]
        loss.backward()
        torch.nn.utils.clip_grad_norm_(train_model.parameters(), 0.5)
        optimizer.step()
        total_loss += loss.item()

    return total_loss / len(batches)

def test(test_model, batches):
    total_loss = 0.
    for i, batch in tqdm(enumerate(batches), total=len(batches)):
        test_model.eval()
        inputs = batch
        loss = test_model(inputs, labels=inputs)[0]
        total_loss += loss.item()

    return total_loss / len(batches)

In [9]:
from torch.optim.lr_scheduler import StepLR

random.shuffle(train_batches)
scheduler = StepLR(optimizer, step_size=2, gamma=0.8)
for epoch in range(10):
    random.shuffle(train_batches)
    loss = train(model, train_batches, optimizer, criterion)
    print('Epoch:', epoch, 'Loss:', loss)
    torch.save({'epoch': epoch,
                'model_state_dict': model.state_dict()},
                'save_small' + str(epoch))
    scheduler.step()

NameError: name 'train_batches' is not defined

In [6]:
_question_prompt = '\nQ: '
_answer_prompt = '\nA: '
    
def get_text_up_to_question_number(text, number):
    pos = text.find(_answer_prompt)
    for _ in range(number):
        pos = text.find(_answer_prompt, pos + 1)
    return text[0:pos + 1]
    
def get_answers_number(text):
    return text.count(_answer_prompt)

def get_answer_number(text, number):
    pos = text.find(_answer_prompt)
    for _ in range(number):
        pos = text.find(_answer_prompt, pos + 1)
    end = text.find('\n', pos + len(_answer_prompt))
    return text[pos + len(_answer_prompt):end]

def get_question_number(text, number):
    pos = text.find(_question_prompt)
    for _ in range(number):
        pos = text.find(_question_prompt, pos + 1)
    end = text.find('\n', pos + len(_question_prompt))
    return text[pos + len(_question_prompt):end]

def get_all_answers(dev_dict, dev_index):
    answers = [[item['input_text'] for item in dev_dict['data'][dev_index]['answers']]]
    answers += [[item['input_text'] for item in dev_dict['data'][dev_index]['additional_answers'][str(index)]] for index in range(3)]
    return [list(set([answers[j][i] for j in range(len(answers))])) for i in range(len(answers[0]))]

In [7]:
def generate_answer_number(model, text, number):
    prompt = get_text_up_to_question_number(text, number)
    tokens = tokenizer.encode(prompt, return_tensors='pt')
    _length = 20
    tokens_length = tokens.shape[1]
    output = model.generate(
             tokens,
             max_length=tokens_length + _length,
             temperature=0,
             pad_token_id=50256
    )
    output = tokenizer.decode(output[0], skip_special_tokens=True)
    return get_answer_number(output, number)

In [8]:
def compute_accuracy_of_model_with_all_answers(model):
    total_number_of_questions = 0
    correct_answers = 0
    wrong_predictions = []

    dlist = dev_list[:10]
    for index, text in tqdm(enumerate(dlist), total=len(dlist)):
        total_questions = get_answers_number(text)
        all_answers = get_all_answers(dev_dict, index)
        for number in range(total_questions):
            prediction = generate_answer_number(model, text, number)
            prediction = prediction.replace('.', '').replace('"', '')
            it_was_answered = False
            for label in all_answers[number]:
                label = label.replace('.', '').replace('"', '')
                if prediction.lower() == label.lower():
                    correct_answers += 1
                    it_was_answered = True
                    break
                elif prediction.lower() in label.lower():
                    correct_answers += 1
                    it_was_answered = True
                    break
                elif label.lower() in prediction.lower():
                    correct_answers += 1
                    it_was_answered = True
                    break
                else:
                    wrong_predictions.append({'label': label, 'prediction': prediction})
            if not it_was_answered:
                print('No Answer for number', number)
            total_number_of_questions += 1

    return correct_answers/total_number_of_questions, wrong_predictions

In [9]:
def get_text_from_data_item(item, max_num_questions=0, question_number=-1, last_question=True):
    text = 'In the text below two people are discussing a story.\n\n'
    text += 'Story:\n' + item['story'] + '\n\n'
    text += 'Discussion:\n'
    text += '\n'.join(['Q: ' + q['input_text'] 
                       + '\nA: ' + a['input_text'] 
                       for q, a in zip(item['questions'][max(0,question_number-max_num_questions):question_number+1], 
                                       item['answers'][max(0,question_number-max_num_questions):question_number+1]) 
                      ])
    if not last_question:
        text = '\n'.join(text.split('\n')[:-1]) + '\n'
    return text

In [10]:
def generate_answer(model, prompt):
    tokens = tokenizer.encode(prompt, return_tensors='pt')
    _length = 50
    tokens_length = tokens.shape[1]
    if tokens_length + _length > 1024:
        return ''
    output = model.generate(
             tokens.cuda(),
             max_length=tokens_length + _length,
             #temperature=0,
             pad_token_id=50256
    )
    output = tokenizer.decode(output[0], skip_special_tokens=True)
    offset = len(prompt)
    start = offset + 1
    end = output.find('\n', start)
    return output[start:end].split(':')[-1].strip()

In [11]:
def generate_multiple_answers(model, prompt, num_replicas=25):
    model.train()
    outputs = []
    with torch.no_grad():
        tokens = tokenizer.encode(prompt, return_tensors='pt')
        tokens = tokens.repeat(num_replicas,1)
        _length = 50
        tokens_length = tokens.shape[1]
        if tokens_length + _length > 1024:
            return ''

        
        output = model.generate(
             tokens.cuda(),
             max_length=tokens_length + _length,
             pad_token_id=50256
        )
        for index in range(num_replicas):
            text = tokenizer.decode(output[index, :], skip_special_tokens=True)
            offset = len(prompt)
            start = offset + 1
            end = text.find('\n', start)
            outputs.append(text[start:end].split(':')[-1].strip())

    return outputs

In [12]:
device = 'cuda'

In [13]:
import numpy as np
from sentence_transformers import SentenceTransformer


sentence_model = SentenceTransformer('msmarco-distilbert-base-v3')
sentence_model = sentence_model.to(device)

In [14]:
def get_embeddings_from_text(text):
    outputs = sentence_model.encode(text)
    return outputs

def group_similar_answers_and_get_scores(answers):
    answers_dict = {}
    threshold = 0.7
    embeddings = get_embeddings_from_text(answers)
    embeddings = np.array([e/np.linalg.norm(e) for e in embeddings])
    similarity_matrix = np.matmul(embeddings, embeddings.transpose())
    superseded = set()
    superseded_from = {}
    for i in range(len(answers)):
        for j in range(len(answers)):
            if i > j:
                continue
            if i != j and answers[i] == answers[j]:
                continue
            if similarity_matrix[i][j] > threshold :
                answers_dict.setdefault(i, 0)
                answers_dict[i] += 1
                if i != j:
                    superseded.add(j)
                    superseded_from.setdefault(i, [])
                    superseded_from[i].append(j)

    answers_and_scores = [(index, score/len(answers))
                          for index, score in answers_dict.items() 
                          if index not in superseded]
    
    new_scores_dict = {}
    total_score = sum(item[1] for item in answers_and_scores)
    for answer_index, score in answers_and_scores:
        answer_group = [answers[answer_index]]
        if answer_index in superseded_from:
            answer_group += [answers[i] for i in superseded_from[answer_index]]
        answer_group = tuple(set(answer_group))
        if answer_group in new_scores_dict:
            new_scores_dict[answer_group] += score / total_score
        else:
            new_scores_dict[answer_group] = score / total_score
    
    
    return sorted(list(new_scores_dict.items()), key=lambda x: -x[1])

In [15]:
group_similar_answers_and_get_scores(['white', 'white', 'white', 'white', 'orange', 'orange', 'white', 'white', 'white', 'white', 'white', 'orange', 'white', 'white', 'White', 'white', 'orange', 'white', 'white', 'white']
)

[(('white', 'White'), 0.846153846153846), (('orange',), 0.15384615384615385)]

In [16]:
small_text = get_text_from_data_item(dev_dict['data'][0], 
                                     max_num_questions=8,
                                     question_number=1,
                                     last_question=False)

answers = generate_multiple_answers(model, small_text)
group_similar_answers_and_get_scores(answers)

[(('in a barn',
   'in the barn',
   'In the barn',
   'in a barn near a farm house.',
   'In the barn.'),
  0.2727272727272727),
 (('',), 0.22727272727272724),
 (('What is Cotton?', 'Cotton'), 0.0909090909090909),
 (('She lived in a barn in a barn.',
   'She lived in a barn in a farm near a farm. farm.'),
  0.0909090909090909),
 (('The farm house in the middle of the woods.',), 0.04545454545454545),
 (("Cotton's mommy",), 0.04545454545454545),
 (('Her mommy and 5 sisters were all white.',), 0.04545454545454545),
 (('Where did she live?',), 0.04545454545454545),
 (('Cotton lived in a nice, warm place above the barn.',),
  0.04545454545454545),
 (('"I was born in a barn in a barn in a barn in a barn in a barn in a barn in a barn in a barn in a barn in a barn in a barn in a barn in a barn in a barn i',),
  0.04545454545454545),
 (("Cotton's home",), 0.04545454545454545)]

In [24]:
def compute_accuracy_of_model(model):
    total_number_of_questions = 0
    correct_answers = 0
    wrong_predictions = []

    false_positives = []
    dlist = dev_list[:100]
    for index, text in tqdm(enumerate(dlist), total=len(dlist)):

        all_answers = get_all_answers(dev_dict, index)
        total_questions = len(all_answers)        
        
        for number in range(total_questions):
            small_text = get_text_from_data_item(dev_dict['data'][index], 
                                                 max_num_questions=8,
                                                 question_number=number,
                                                 last_question=False)
            predictions = generate_multiple_answers(model, small_text)
            predictions = group_similar_answers_and_get_scores(predictions)
            all_predictions = []
            #for prediction in predictions[:2]:
            #    all_predictions += prediction[0]
            best_prediction_score = predictions[0][1]
            for prediction in predictions:
                if prediction[1]/best_prediction_score < 0.8:
                    continue
                all_predictions += prediction[0]
            print(all_predictions)
            for prediction in all_predictions:
                it_was_answered = False
                if not prediction:
                    print('NO PREDICTION!!')
                    prediction = 'unknown'
                prediction = prediction.replace('.', '').replace('"', '')
                it_was_answered = False
                for label in all_answers[number]:
                    label = label.replace('.', '').replace('"', '')

                    if prediction.lower() != 'unknown' and label.lower() == 'unknown':
                        false_positives.append(prediction)

                    if prediction.lower() == label.lower():
                        correct_answers += 1
                        it_was_answered = True
                        break
                    elif prediction.lower() in label.lower():
                        correct_answers += 1
                        it_was_answered = True
                        break
                    elif label.lower() in prediction.lower():
                        correct_answers += 1
                        it_was_answered = True
                        break
                    else:
                        wrong_predictions.append({'label': label, 'prediction': prediction})
                        
                if it_was_answered:
                    break

            total_number_of_questions += 1

    return correct_answers/total_number_of_questions, wrong_predictions, false_positives

In [25]:
false_positives = None
wrong_answers = None


checkpoint = torch.load('save_adversarial' + str(6))
model.load_state_dict(checkpoint['model_state_dict'])
accuracy, wrong_answers, false_positives = compute_accuracy_of_model(model.cuda())
print('Epoch:', 6, ' with accuracy:', accuracy)

  0%|          | 0/100 [00:00<?, ?it/s]

['white', 'White']
['in a barn', 'a barn']
['no']
['her mom and 5 other sisters', 'her mommy and 5 other sisters']
['orange']
['no']
['she used her fur to paint herself', 'used it to paint herself', 'she painted herself', 'she used it to paint herself', 'painted herself like them', 'painted herself', 'she used her paint', 'she used her paint to paint herself', 'she painted herself like them', 'used her paint to paint herself']
["her mommy's", 'her mommy']
['laughed', 'they laughed']
['in a bucket of water', 'a bucket of water', 'a bucket']
['licked her face', 'they licked her face']


  1%|          | 1/100 [00:12<20:21, 12.33s/it]

['yes']
['Asta']
['The bottle', 'The bottle.', 'A bottle']
["Asta's friend"]
['yes', 'Yes']
['Yes']
['a note', 'A note']
['Yes']
['Asta']
["they took it to Asta's papa", "They took it to Asta's papa.", 'They took it to the papa', 'they took it to his papa', "took it to Asta's papa", "They took it to Asta's papa"]
['Yes']


  2%|▏         | 2/100 [00:22<18:01, 11.04s/it]

['yes', 'Yes']
['a Chinese lady', 'The elderly Chinese lady', 'the elderly Chinese lady', 'a lady']
['yes', 'Yes']
['A paper carrier bag', 'a paper carrier bag', 'Paper carrier bag']
['Yes']
['Nicole']
['Shanghai']
['She is his grandmother', 'Her mother is his grandmother', "He's her grandmother", 'His grandmother', 'He is his grandmother']
['a thermos with hot soup and a stainless-steel container with rice, vegetables and either chicken, meat or shrimp', 'a thermos with hot soup and a stainless-steel container with rice, vegetables and chicken, meat or shrimp, sometimes with a kind of pancake', 'a thermos with hot soup and a stainless-steel container with rice', 'a thermos with hot soup and a stainless-steel container with rice, vegetables and either chicken, meat or shrimp, sometimes with a kind of pancake.', 'a thermos with hot soup and a stainless-steel container with rice, vegetables and chicken, meat or shrimp', 'a thermos with hot soup and a stainless-steel container with rice, 

  3%|▎         | 3/100 [00:41<23:49, 14.74s/it]

['Thank you', 'thank you', 'Thank you"']
['yes']
['Dennis Farina']
['dapper']
['No.', 'No']
['Yes', 'Yes.']
['Yes', 'Yes.']
['He joined the cast of Law & Order', 'He was on "Law & Order"', 'he joined the cast of the long-running "Law & Order"', 'He joined the cast of the long-running "Law & Order"']
['Michael Mann']
['"Thief"', 'Thief', '"Thief."']
['Police or gangsters', 'Police officers or gangsters', 'Police or Gangster', 'Police or gangster', 'cops or gangsters', 'Police and gangsters', 'Police officer']
['he joined the cast of the long-running "Law & Order"', 'He joined the cast of the long-running "Law & Order"', 'joined the cast of the long-running "Law & Order"']
['Law & Order', '"Law & Order"']
['Jerry Orbach', 'Joe Fontana', 'Detective Joe Fontana']
['No']
['expensive car', 'a expensive car', 'a cheap car', 'a car']
['No']
['flashy', 'flashy clothes and an expensive car', 'They were expensive']
['No']
['Yes']


  4%|▍         | 4/100 [01:04<28:40, 17.92s/it]

['A cop', 'a cop', 'Police officer']
['to and from school']
['yes']
['walk in and to the bus stop', 'walk home from the bus stop', 'walk to the bus stop']
['No']
['Yes']
['story time']
['right before bedtime', 'before bedtime']
['she thought something might be wrong']
['no', 'No']
['No']
['that she was upset', 'she was upset']
['yes', 'Yes']
['that everything would be okay', 'that he was sure everything would be okay', 'that he was sure that everything would be okay', 'he was sure that everything would be okay']
['her teacher']
['no', 'No']
['the dentist', "Quinton's mother", 'Quinton', 'her mother']
['the dentist', 'to the dentist']


  5%|▌         | 5/100 [01:23<29:15, 18.48s/it]

['tomorrow']
['Five']
['New York', 'New York City']
['New York']
['yes']
['Staten Island']
['Conference House Park']
['476,013', '476,010', '476,015']
['No']
['white', 'White']
['the forgotten borough', '"the forgotten borough"', 'forgotten borough', 'The forgotten borough']
['by inhabitants feel neglected by the city government', 'by inhabitants who feel neglected by the city government', 'by residents who feel neglected by the city government', 'because the city government feels neglected', 'by residents who feel neglected', 'residents feel neglected', 'residents feel neglected by the city government']
['the East Shore', 'East Shore']


  6%|▌         | 6/100 [01:37<26:15, 16.76s/it]

['St George, Tompkinsville, Clifton, and Stapleton', 'St. George, Tompkinsville, Clifton, and Stapleton', 'St. George, Tompkinsville, Clifton, and St. George Historic District, which feature large Victorian houses']
['the morning', 'five in the morning', 'at five in the morning']
['the weather forecast', 'weather', 'the weather']
['yes', 'Yes']
['firefighter', 'Fireman', 'Firefighter']
['Yes']
['The flashlight', 'flashlights', 'A flashlight', 'Flashlights', 'flashlight']
['R.J.']
['Joel', 'His son']
['Glass, wood, and plaster']
['Yes']
['R.J.']


  7%|▋         | 7/100 [01:52<25:19, 16.33s/it]

["The light of his father's flashlight", 'The light of his flashlight']
['Gary Giordano', 'Giordano']
['Gaithersburg']
['Montgomery County']
['Alabama']
['Aruban jail', 'in an Aruban jail']
['he is being held in an jail', 'he is being held in an Aruba jail', 'He is being held in an Aruban jail', 'he is currently being held in an Aruban jail', 'he is being held in an Aruban jail']
['FBI']
['15']
['Philip Celestini', 'Taco Stein']
['Monday']
['eight more days']
['Robyn Gardner']
['Baby Beach', 'near Baby Beach', 'baby beach']
['swimming']
['Robyn Gardner', 'with Gardner', 'Gardner']
['no']
['no', 'No']
['50']
['August 5,', 'August 5']


  8%|▊         | 8/100 [02:17<29:05, 18.97s/it]

['unknown']
['India']
['India']
['30 feet tall', '30 feet', 'Three or four feet tall', 'three or four feet', 'Three feet', 'Three or four feet tall.', 'Three or four feet']
['they lost nearly all of its healthy qualities', 'it loses nearly all of its healthy qualities', 'They lose nearly all of its healthy qualities', 'they lose nearly all of its healthy qualities', 'it lost nearly all its healthy qualities', 'it lost all its healthy qualities', 'it loses nearly all its healthy qualities', 'it loses all of its healthy qualities']
['prevents heart disease', 'prevent heart disease', 'it prevents heart disease']
['by accident.', 'by accident']
['Shen Nong']
['about 2737 B.C', '2737 B.C', '2737', '2737 B.C.']
['yes']


  9%|▉         | 9/100 [02:29<25:19, 16.70s/it]

['he drank several cups of hot water daily', 'he drank several cups of hot water']
['Der Spiegel']
['Germany']
['posing over the bodies of dead Afghans', 'posing over dead Afghans']
['sprawled over a patch of sand and grass', 'it was sprawled over a patch of sand', 'a bloody body sprawled over a patch of sand and grass', 'it was laid over a patch of sand and grass', 'it was laid on a patch of sand and grass', 'Sprawled over a patch of sand and grass']
['soldiers kneeling by a bloody body', 'soldiers kneeling by a bloody body sprawled over a patch of sand', 'soldiers kneeling by a bloody body sprawled over a patch of sand and grass.', 'soldiers kneeling by a bloody body sprawled over a patch of sand and grass']
['a post in front of a military vehicle']
['weapons', 'a weapon']
['Jeremy Morlock']
['Pfc. Andrew Holmes', 'Andrew Holmes']


 10%|█         | 10/100 [02:39<22:00, 14.67s/it]

['murder', 'the premeditated deaths of three civilians', 'premeditated deaths of three civilians', 'premeditated deaths']
['Mayweather and Manny Pacquiao', 'Mayweather and Pacquiao', 'Floyd Mayweather and Manny Pacquiao']
['The Money Man', '"The Money Man"']
['TBE']
['The Money Man']
['The Money Team']
['CEO of Team Sauerland']
['5.65 million', 'nearly five million', 'Five million']
['$300 million']
['47', '47-0']


 11%|█         | 11/100 [02:50<20:16, 13.67s/it]

["they're both playing their roles", "they're both playing roles"]
['OCLC Online Computer Library Center, Incorporated', 'OCLC Online Computer Library Library Center', 'OCLC Online Computer Library Center']
['OCLC Online Library Center', 'OCLC Online Computer Library Center']
['1967']
['no']
['Ohio College Library Center']
['the Ohio College Library Center', 'Ohio College Library Center', 'College Library Center']
['Frederick G. Kilgour']
['Yes']
['Librarian', 'librarian', 'He is a librarian']
['WorldCat']
['July 5, 1967']
['Ohio State University', 'on the campus of the Ohio State University', 'the campus of the Ohio State University', 'campus of the Ohio State University', 'the Ohio State University']
['the Alden Library', 'Alden Library']
['Ohio University']
['streamline operations, control costs, and increase efficiency', 'streamline operations']
['August 26, 1971']


 12%|█▏        | 12/100 [03:08<22:01, 15.02s/it]

['yes', 'Yes']
['yes', 'Yes']
['a sweater', 'A sweater']
['$15', '$15.', '$10']
['No']
["It's only $10.", "It's only $15.", 'It looks like diamonds', 'It looks like diamonds.']
['Summer.', 'summer', 'Summer']
['$15', '$15.']
['No']
['A pen.', 'A pen', 'Pen', 'Pen and pencils']
["It's only $10.", "It's only $15.", "It's $15", "It's only $15", "It's only $10"]
['summer', 'Summer']


 13%|█▎        | 13/100 [03:19<19:53, 13.72s/it]

['$15']
['by a big lake']
['two little mice']
['lots of toy boats', 'toy boats', 'a toy boat', 'a big toy boat']
['yes']
['little mouse and Mary', 'little mice', 'Mary, one of the little mice', 'little mice', 'mice']
['the water', 'from the water']
['looking at the water', 'look at the water']
['swimming', 'swimming and splashing in the water', 'swimming and splashing', 'splashing in the water']
['threw a ball', 'threw a ball into the water']


 14%|█▍        | 14/100 [03:29<18:17, 12.76s/it]

['landed by the toy boat', 'it landed by the toy boat']
['He had been bombed', 'It had been bombed', 'it had been bombed', 'he had been bombed', 'It was bombed', 'bombed']
['No.', 'No']
['Germany.', 'Germany']
['Eastern', 'Eastern.']
['Western Germany', 'Western Germany.']
['No.']
['No.', 'No']
['No.', 'No']
['Yes', 'Yes.', 'No.', 'No']
['Return to his home.', 'Go back into his home.', 'He returned to his home.', 'He returned to his home', 'Go back to his home.']
['Yes']
['A taxi-driver', 'taxi-driver']
['Yes']
['Hans Bussman', 'Franz Bussman']
['No']
['No', 'Yes']
['He was thought to have been killed in action', 'He was thought to have been killed in action.', 'He was thought to have been killed', 'He was thought to have been killed twenty years ago.', 'He was thought to have been killed.']
['Mrs. Bussman', 'Mrs. Busman', 'Mrs. Busman and her husband.']
['He might be his brother.', 'He thought there was a chance that he might be his brother', 'He thought he might be his brother.', 'He

 15%|█▌        | 15/100 [03:50<21:27, 15.15s/it]

['No']
['ard']
['riel', 'riel_']
['What was the name of the island?', 'What was the name of the vessel?', 'What was the name of the ship?']
['s']
['What color was it?']
['What was the new name for himself?', 'What was the name of the new name?', 'What was the new name for him?', 'What was the new name?']
['a', 'eward', 'eward coast', 'eward coast of Malaita']
['']
NO PREDICTION!!
['sband']
['', 'rangi_.', 'rangi_', 'rangi']
NO PREDICTION!!
['back to Tulagi', 'to Tulagi', 'we get back to Tulagi']
['sband']
['What was the name of the ship that was named after the boat?', 'What was the name of the boat that was in the lagoon?', 'What was the name of the boat?']
["Riggs,'", 'Riggs', "Riggs,' or 'Mademoiselle de Maupin,' or 'Judas,' or 'Haman.'"]
[".'", "oiselle de Maupin,'", "oiselle de Maupin'", "oiselle de Maupin,' or 'Judas,'"]
["'", ",'", ".'"]
['What was the name of the boat that was on the lagoon?', 'What was the name of the ship that was in the water?', 'What was the name of the boa

 16%|█▌        | 16/100 [04:16<25:37, 18.31s/it]

['What was the name of the ship that was in the lagoon?', 'What was the name of the ship that was in the sea?']
['Brownie and Spotty']
['every day']
['loved each other', 'they loved each other']
['between their houses', 'their respective houses', 'between their respective houses']
['neighbor dog', 'neighbor dogs']
['Brownie']
['yes', 'Yes']
['yes']
['a deserted spot half a mile from the house', 'a deserted spot']
['he followed Ted around', 'he followed Ted about', 'He followed Ted']
['led him to a nearby empty lot and back', 'led him to a nearby empty lot', 'ran toward a nearby empty lot and back']
['keep his spirits up', 'kept his spirits up']
['yes', 'Yes']
['a path through the field between their houses', 'a path through the grass of the field', 'a path through the grass of the field between their houses']
['yes', 'Yes']
['Yes']
['no', 'No']
['yes', 'Yes']
['no', 'No']
['he was still missing', 'he was missing']


 17%|█▋        | 17/100 [04:41<28:02, 20.27s/it]

["Follow me! It's urgent!", '"It\'s urgent!"', '"Follow me! It\'s urgent!"']
['A young girl and her dog', 'a young girl and her dog', 'a girl and her dog']
['walking.', 'They were walking', 'They were walking.', 'walking']
['Into the woods.', 'Into the woods']
['a little scared', 'scared', 'She was scared']
['very interested', 'interested']
['very interested', 'interested']
['In the bushes', 'What was in the bushes up ahead', 'What was in the bushes', 'in the bushes', 'bushes', 'what was in the bushes', 'the bushes', 'in what was in the bushes']
['a brown bear', 'a small brown bear']
['rested his head on his paws', 'He rested his head on his bear paws', 'rested his head on his bear paws']
['yes', 'Yes']
['very interested']
['He then rested his head on his bear paws and went back to sleep', 'He rested his head on his bear paws']
['rested his head on his paws', 'He rested his head on his bear paws', 'rested his head on his bear paws', 'He then rested his head on his bear paws', 'he reste

 18%|█▊        | 18/100 [04:57<25:56, 18.98s/it]

['yes', 'Yes']
['a 10-year-old boy fatally shot his father', 'A 10-year-old boy fatally shot his father', 'a boy fatally shot his father']
["his mother's home"]
['no']
["in the back seat of the man's Toyota 4-Runner"]
['when the shooting took place', 'when the shooting occurred', 'when the shooting occured']
['3pm', 'about 3pm']
['Rick James Lohstroh']
['exited the back of the vehicle and continued to fire at the car', 'exited the back of the vehicle', 'he exited the back of the vehicle and continued to fire at the car', 'he exited the back of the vehicle', 'he exited the back seat of the vehicle']
['no']
['the University of Texas Medical Branch', 'University of Texas Medical Branch', 'at the University of Texas Medical Branch']
['Memorial Herman Hospital']
['Yes']


 19%|█▉        | 19/100 [05:11<23:53, 17.69s/it]

['a 7-year-old', 'the 7-year-old', "the boy's mother"]
['Romola']
['no', 'No']
['no']
['he was a clergyman', 'he was a priest', 'he was a heretic']
['Romola']
['Rome', 'in Rome']
['in June', 'June']
['a week', 'weeks']
['the scene', 'Savonarola']
['a conspiracy', 'the conspiracy', 'a revelation of the suspected conspiracy', 'a sign from Savonarola']
['the lengthening sunny days', 'lengthening sunny days']
['the Frate']


 20%|██        | 20/100 [05:26<22:14, 16.69s/it]

['yes', 'Yes']
['yes', 'Yes']
['to Utah', 'Utah']
['in Utah', 'Utah']
['no']
['a small apartment']
['no']
['her friends']
['yes']
['Three', 'three', 'Four']
['the truck', 'a truck']
['her dad']
['yes']
['yummy fast food', 'yummy']
['yes']
['a knock', 'a knock at the door']
['a little girl']
['to play']


 21%|██        | 21/100 [05:40<20:55, 15.89s/it]

['she loved it', 'sad']
['she fell on a ski slope', 'she fell']
['Ski', 'Skiing', 'skiing']
['Canada.', 'Canada']
['Yes', 'Yes.']
['Yes.']
['About an hour.']
['No.']
['The Lenox Hill Hospital.', 'Lenox Hill Hospital.']
['New York City.']
['Lenox Hill Hospital.']
['45.']
['Actress.', 'Actor', 'Actress', 'A stage actress']
['No.', 'No']
['Liam Neeson']
['Yes.']
['Boys.', 'Boys']
['Yes.']
['Tony-winning stage actress']
['Yes.']


 22%|██▏       | 22/100 [06:04<24:06, 18.54s/it]

['She was a star', 'She was a film star', 'She was a star.', 'She was a stage actress.', 'As a stage actress', 'She was a stage actress']
['His bark', 'his bark']
['three months']
['no']
['Sammie']
['a golden puppy', 'golden puppy']
['no']
['he was sad', 'he was very sad', 'he was very sad and tired']
['yes']
['Peter']
['find a person']
['no']


 23%|██▎       | 23/100 [06:14<20:11, 15.73s/it]

['held him in her arms', 'he held her']
['a cute alien dog', 'a cute dog']
['false', 'False']
['Three']
['Ti, Dicky and their dog CJ7', 'Ti, his son, and CJ7', 'Dicky, Ti, and CJ7', 'Dicky, Ti and CJ7', 'Ti, Dicky and CJ7', 'Dicky, Dicky and CJ7']
['no']
['in the movie', 'in a movie']
['no']
['China']
['Dicky']
['no']
['a doll', 'doll']
['false']
['in the trash', 'in a trash']
['no']
['Talk and do magic', 'talk and do magic']
['o']
['worker']
['anuary']


 24%|██▍       | 24/100 [06:29<19:56, 15.74s/it]

['008']
['art and crafts', 'art-and-crafts', 'art and craft']
['recipe', 'recipes']
['Heather Neroy', 'Neroy']
['Southern California']
['a stay-at-home mom', 'stay-at-home mom', 'stay at home mom', 'mom']
['copied the link', 'copying the link', 'by copying the link', 'she copied the link']
['e-mail', 'an e-mail']
['no']
['a new website called Pinterest', 'a website called Pinterest', 'Pinterest', 'a website calledPinterest']
['no']
['yes']
['yes']
['a Halloween board', 'Halloween board']
['a shared color board', 'yes']
["redecorating her daughter's bedroom", "her daughter's bedroom"]
["follow each other's boards"]
['re-pin']
['yes']
['neat to see what other moms are pinning', '"It\'s been really neat to see what other moms are pinning', "it's been really neat to see what other moms are pinning"]


 25%|██▌       | 25/100 [06:53<22:26, 17.96s/it]

['yes']
['Lord Earl', 'Sir Earl']
['his traces', 'traces']
['Get forward.', 'Get forward,', 'Get forward', 'get forward', 'to get forward']
['No.', 'No']
['orders to prepare to start with the hound', 'To prepare to start with the hound', 'Orders to prepare to start with the hound']
['Three']
['To prepare to start with the hound', 'prepare to start with the hound', 'Prepare to start with the hound', 'Start with the hound', 'Preparing to start with the hound', 'To start with the hound']
['The earl and a large party of men-at-arms', 'A large party of men-at-arms']
['A traitor', 'The traitor']
['Where Bruce had slept', 'Where Bruce had slept the night before.', 'Where Bruce had slept the night before', 'where Bruce had slept', 'That Bruce had slept the night before']
["He was determined that if under the dog's guidance the party came close up with Bruce, he would kill the dog and then try to escape by fleetness of foot", "He was determined that if under the dog's guidance the party came cl

 26%|██▌       | 26/100 [07:15<23:39, 19.18s/it]

['Pembroke and his knights']
['Bartolome Ramon Pineda Ortiz', 'Brig. Gen. Bartolome Ramon Pineda Ortiz', 'Cibar Benitez', "the president's office"]
['Paraguay']
['yes']
['Fernando Lugo', 'President Fernando Lugo']
['the top brass']
['Bartolome Ramon Pineda Ortiz', 'Brig. Gen. Bartolome Ramon Pineda Ortiz']
['yes']
['Brig. Gen. Hugo Gilberto Aranda Chamorro and Rear Adm. Egberto Emerito Orie Benegas took over the top posts at the air force and navy', 'Brig. Gen. Hugo Gilberto Aranda Benegas', 'Brig. Gen. Hugo Gilberto Aranda Chamorro', 'Brig. Gen. Hugo Gilberto Aranda Chamorro and Rear Adm. Egberto Orie Benegas', 'Brig. Gen. Hugo Gilberto Aranda Chamorro and Rear Adm. Egberto Emerito Orie Benegas', 'Hugo Gilberto Aranda Chamorro']
['Brig. Gen. Hugo Gilberto Aranda Chamorro', 'Hugo Gilberto Aranda Benegas', 'Hugo Gilberto Aranda Chamorro', 'Brig. Gen. Hugo Gilberto Aranda']
['yes']
['Benitez', 'Bartolome Benitez']
['1989']
['yes']
['1999 and 2000', '1996 and 2000', '2000 and 1999', '2000

 27%|██▋       | 27/100 [07:40<25:37, 21.06s/it]

['he has struggled to push reforms through a majority-opposition legislature', 'he has tried to push reforms through a majority-opposition legislature', 'he has struggled to push reforms through a majority-oppopposition legislature', 'he has struggled to push through reforms through a majority-opposition legislature', 'he has struggled to push reforms through a majority-opposition legislature.', 'has struggled to push reforms through a majority-opposition legislature', 'struggled to push reforms through a majority-opposition']
['a dress']
['to her party', 'to wear to her party']
['es']
['er dad', 'er mom', 'er parents']
['']
NO PREDICTION!!
['ug']
['in the morning']
['se it was opposite day', 's opposite day']
['what did they buy instead?', 'what did they buy?']


 28%|██▊       | 28/100 [07:47<20:21, 16.97s/it]

['steaks', 'what did they get a milkshake for?', 'vanilla milkshake and went to more stores', 'vanilla milkshake', 'did they get a vanilla milkshake?']
['Buckinghamshire']
['south east', 'South East England', 'in South East England', 'South East']
['Greater London']
['Oxfordshire', 'Berkshire']
['Northamptonshire']
['Northamptonshire']
['Northamptonshire']
['Aylesbury']
['home counties']
['no', 'No']
['the Metropolitan Green Belt', 'The Metropolitan Green Belt.', 'Metropolitan Green Belt', 'The Metropolitan Green Belt']
['the Metropolitan Green Belt', 'Metropolitan Green Belt']
['Yes', 'Yes.']
['Milton Keynes']
['the northeast', 'in the northeast']


 29%|██▉       | 29/100 [08:02<19:15, 16.27s/it]

['the Conservative Party', 'The Conservative Party', 'Conservative Party', 'Conservative']
['Pierre']
['yes']
['to look after the horses']
['twelve miles away', 'twelve miles', '12 miles away']
['four hours']
['no']
['The landlady', 'the landlady']
['yes']
['no']
['it was very poor', 'very poor']
['yes']
['yes']
['yes']
['a party']
['no']
['yes']
['yes']
['that he wished to speak to him', 'that he had spoken to him']
['an honest man', 'he is an honest man', 'honest']


 30%|███       | 30/100 [08:23<20:46, 17.81s/it]

['Pierre']
['high upland and wide valleys', 'high downland and wide valleys', 'its high downland and wide valleys']
['in South West England', 'South West England']
['no']
['the counties of Dorset, Somerset, Hampshire, Gloucestershire, Oxfordshire and Berkshire', 'Dorset, Somerset, Hampshire, Gloucestershire, and Oxfordshire', 'Dorset, Somerset, Hampshire, Gloucestershire, Oxfordshire and Berkshire', 'Dorset, Somerset, Hampshire, Gloucestershire, Berkshire']
['Wilton', 'Wiltonshire']
['Trowbridge']
['its mediaeval cathedral', 'it is famous for its mediaeval cathedral', 'it is notable for its mediaeval cathedral', 'its mediaeval cathedral.', 'for its mediaeval cathedral']
['yes']
['Longleat']
['Warminster']
['Stourhead', "the National Trust's Stourhead", 'Stourh']
['near Mere', 'Mere', 'near Mere.']
['as "Wiltunscir"', 'Wiltunscir', '"Wiltunscir"']
['Pre-Roman', 'pre-Roman archaeology', 'pre-Roman']
['Stonehenge and Avebury stone circles', 'Stonehenge and Avebury']
['the Battle of Bedwyn

 31%|███       | 31/100 [08:47<22:32, 19.60s/it]

['large areas of the country came into the possession of the crown and the church', 'large areas of the country came into the possession of the crown', 'large areas of the country came into the possession of the crown and the church.']
['Male', 'male', 'a male']
['A male.']
['Marsha.', 'Marsha']
['Her mom', 'Her mom.']
['To get him safe and sound at all times.', 'To make sure Joey was safe and sound at all times.', 'To be safe and sound at all times.', 'To make sure Joey is safe and sound at all times.']
['Tuesday.', 'On one Tuesday.', 'On Tuesday.']
["Marsha's favorite dinner was spaghetti", "Marsha's favorite dinner.", "Marsha's favorite dinner is spaghetti.", "Marsha's favorite dinner was spaghetti."]
['No.']
['He gets a little dried out.', 'He gets dried out.', 'Every few days.']


 32%|███▏      | 32/100 [08:55<18:13, 16.08s/it]

['Yes.']
['Jane Humphreys', 'Jane Humphrey']
['a book of Parthenissa', 'Parthenissa', 'a volume of Parthenissa']
['disappointment']
['Portchester', 'at Portchester']
['no']
['disappointment']
['with the guests of royalty', 'no']
['to her companions', 'her companions or anything around her', 'her companions or', 'her companions', 'to her companions or anything around her']
['Hester', 'Hester Bridgeman']
['no']
['no']
['Three', 'three']
['no']
['an attorney']
['no']


 33%|███▎      | 33/100 [09:15<19:06, 17.11s/it]

['MACHINATIONS']
['Tom Rover']
['Stanley']
['Stanley', 'Sam']
['"What did you do to him?"', '"What did he do to him?', '"What did he do to him?"', 'What did he do to him?']
['put an advertisement of pills on his back', 'put an advertisement of pills on his back and some other ads', 'put an advertisement on his back', 'Put an advertisement of pills on his back', 'put an advertisement of pills on his back and some other ads.', 'Put an advertisement of pills on his back and some other ads.']
['Put an ad on his back', 'An advertisement of pills on his back', 'put an advertisement of pills on his back', 'An advertisement of pills', 'Put an advertisement of pills on his back']
['yes', 'Yes']
['yes']
['shaking his other fist', 'shaking his fist']
['shaking his fist wrathfully', 'shaking his other fist', 'holding his fist', 'shaking his fist', 'holding the poster Tom had fastened on his back', 'holding the poster', 'shaking his fist wrathfully', 'shaking his other fist', 'shaking his other fis

 34%|███▍      | 34/100 [09:43<22:36, 20.55s/it]

['no']
['cats', 'Cats']
['8', '7']
['Male']
['1']
['Brendan', 'cats']
['special treats']
['Three', 'three', '3']
['because he likes them', 'because he loves them']
['chips and cake', 'chips and cake and candy']
["they aren't good for cats", "those foods aren't good for cats", "because they're not good for cats", "those aren't good for cats", 'those are not good for cats', "because those foods aren't good for cats", "Because they're not good for cats", "because they aren't good for cats"]
['ball of paper', 'chase balls of paper', 'balls of paper']
['Brendan']
['orange, black, and white', 'orange and black', 'white and black', 'orange, black, and spotted', 'white, black, and red']
['Snowball', 'snowball']
['no', 'No']
['female']
['Snowball']
['yes']


 35%|███▌      | 35/100 [09:58<20:23, 18.82s/it]

['Scruff', 'Scruffy']
['a manuscript']
['wiki-wiki', '"Wik-wiki."', 'A wiki-wiki', "'Wik-wiki.'", 'Wik-wiki', "'Wiki-wiki.'"]
['no', 'No']
['In the Hawaiian Islands', 'Hawaii', 'In Hawaii']
['Three', 'Four', 'Two']
['No']
["It's true"]
['Maria']
['To see if he would be at their table', "To see if he'd be at their table", 'To see if he was at dinner', "To make sure he wouldn't be at the table", 'To see if he was at the table', 'To make certain whether or not he would be at their table']
['No']
['Yes']
['Yes']
['Martin']
['at the gate', 'At the gate']
['Maria']
['No']
['Martin']
['No']
['No']


 36%|███▌      | 36/100 [10:20<20:57, 19.65s/it]

["It's too strong", "Because it's too strong"]
['from his home in Dubai', 'his home in Dubai', 'McLaren']
['because of a high impact', 'from his home in Dubai', 'he was concussed and airlifted to hospital']
['high speed', 'high speed crash']
['February 22']
["Barcelona's Circuit de Catalunya"]
['the McLaren simulator', 'McLaren simulator', 'McLaren']
['a comeback', 'his comeback', 'comeback']
['Lewis Hamilton and his mighty Mercedes', 'Lewis Hamilton and his Mercedes']
['world champion']


 37%|███▋      | 37/100 [10:30<17:45, 16.91s/it]

['McLaren']
['Ben and Johnny']
['he was suspicious', 'he was a suspicious sort of a person', 'that he was a suspicious sort', 'that he was a suspicious sort of a person']
['a suspicious sort of a person', 'a sort of a suspicious sort of a person']
['Boarding-house', 'a regular boarding-house', "Master Spry's misfortune"]
['yes']
['five to ten dollars per week', 'five dollars per week']
['ten dollars per week', 'three dollars per week', 'at various sums from five to ten dollars per week', '10 dollars per week', 'five to ten dollars per week']
["Master Spry's misfortune", "Master Spry's misfortune and Tim's perfidy", "Master Spry's misfortune and Tim Dooley's perfidy"]
['yes']
['they were to start a regular theatre', 'that they were to start a regular theatre, and had already engaged a hall', 'that Johnny had told them that they were to start a regular theatre', 'they were to start a regular theatre, and had already engaged a hall', 'that Johnny himself had told him that they were to sta

 38%|███▊      | 38/100 [10:54<19:35, 18.96s/it]

['It was a rumor', 'it was not a rumor', 'it would be regarded as a rumor', 'it was regarded as a rumor', 'it would have been regarded as a rumor', 'it would be regarded simply as a rumor', 'it was a rumor']
['basketball', 'Basketball']
['the NBA players', 'NBA players', 'the NBA players they watch on TV']
["Jared's"]
['Jared', 'Adam']
['they get to play on a real team', 'They get to play on a basketball team', 'They all get to play on a real team', 'they all get to play on a basketball team', 'they play on a basketball team', 'They play on a basketball team', 'they get to play on a basketball team']
['at their school', 'their school']
['practicing a lot', 'practicing']
['every day', 'almost every day']
['Adam', 'Ryan']
['birthday', 'his birthday']


 39%|███▉      | 39/100 [11:02<16:05, 15.83s/it]

['Alex and Brady']
['no', 'No']
['They were too yellow', 'They were too yellow.']
['Green.', 'Green']
['Yes', 'Yes.']
['Yes.']
['Clovers and leaves and vines.', 'Grass and clovers and vines.', 'Grass and clovers.', 'grass and clovers.', 'Grass and clovers and leaves.', 'Grass and clovers and leaves and vines.']
['Sean', 'Sean.']
['No.']
['No.']
['Zarah and Zarah.', 'Zarah.', 'Zarah']


 40%|████      | 40/100 [11:13<14:12, 14.20s/it]

['A piece of green in a necklace.', 'A piece of green in a golden necklace.']
['Miss Baldwin', 'Miss Baldwin Wingate']
['Miss Baldwin', 'Baldwin']
['no', 'No']
['On the following evening', 'on the following evening', "after eleven o'clock on the following evening"]
['yes']
['No']
['Josephine']
['easy-chair', 'Easy-chair', 'a easy-chair']
['Domestic ructions', 'domestic ructions']
['After dinner', 'After breakfast', 'after dinner']
['Yes']
['Wilshaw']
['half a sovereign every time I leave the cab', 'at least half a sovereign every time I leave the cab', 'half a sovereign every time he leaves the cab', 'at least a sovereign every time I leave the cab', 'half a sovereign', 'a quarter a sovereign every time', 'a sovereign every time', 'half a sovereign every time']
['when he leaves the cab', 'every time I leave the cab', 'when I leave the cab', 'every time']
['yes', 'Yes']
['Jimmy']
['anywhere']


 41%|████      | 41/100 [11:34<16:01, 16.30s/it]

['sandwiches']
['2013']
['Pope Emeritus']
['April', 'April 27', 'April 16', '16 April']
['Aloisius Ratzinger', 'Joseph Aloisius', 'Joseph Aloisius Ratzinger']
['Pope Benedict', 'Pope Benedict XVI']
['professor', 'a professor', 'priest']
['no', 'No']
['professor', 'a professor']
['five']
['Bavaria']
['John Paul II', 'Pope John Paul II']
['1926', '1927']


 42%|████▏     | 42/100 [11:46<14:24, 14.91s/it]

['Aloisius Ratzinger', 'Ratzinger']
["at Aaron Poole's home", "Aaron Poole's home"]
['he was in his right mind', 'He was in his right mind then']
['Three']
['Nat Poole, Dad, and Dave', 'Nat Poole, Dad', 'Nat Poole, Dave and Dave', 'Dave, Nat Poole and Laura', 'Nat Poole and Dave', 'Nat, Dave and Dave', 'Nat Poole, Dave, and Laura', 'Dave, Laura, and Roger', 'Dave and Roger', 'Dave, Roger, and Laura', 'Dave, Ben, Roger']
['no', 'No']
['no']
['telegram', 'his telegram']
['Aaron', 'to Aaron']
['set off on a hunt', 'set off a hunt for the wild man', 'set off on a hunt for the wild man', 'they set off on a hunt for the wild man']
['for the wild man', 'the wild man']
['no']
['yes']
['a telegram']
['tomorrow']
['Dave']
["he's tired out"]
['yes']
["they'll have to give up", "they'll have to give it up", 'give up']
['no']


 43%|████▎     | 43/100 [12:08<16:16, 17.14s/it]

['Bear Camp', 'see Bear Camp', 'get to Bear Camp']
['information from various perspectives', 'information']
['transition', 'transition period', 'transition period', 'transitional period', 'a transitional period']
['the preparation of children for adult roles', 'to prepare children for adult roles', 'preparation of children for adult roles']
['transition']
['hormones']
["a deeper voice and larger adam's apple", "a deeper voice and larger adam's apple in boys", 'a deeper voice and larger apple', 'a deep voice and larger apple in boys', 'a deeper voice and larger apple in boys']
['hormones']
['the hormones', 'hormones']
['hormones', 'the hormones']
['a chain reaction', 'the chain reaction']
['tissue responsiveness', 'tissue responsiveness or obesity']


 44%|████▍     | 44/100 [12:19<14:23, 15.41s/it]

['hormones', 'the hormones', 'testes', 'the testes']
['Paul', 'Chet Winthrop']
['talking']
['two', 'Two']
['Barnaby Winthrop']
['They were moving against Captain Grady', 'moving against Captain Grady', 'Stealing', 'stealing']
["he was a member of the horse thieves' gang", 'he was a horse thief', "he was a colored member of the horse thieves' gang", 'the words uttered by Jeff Jones', 'they were interested in the words uttered by Jeff Jones', 'They were interested in the words uttered by Jeff Jones', 'he was interested in the words uttered by Jeff Jones', 'in the words uttered by Jeff Jones']
['yes']
['he was a dirty horse thief', 'he was a thief', 'he was a downright horse thief', 'he was a horse thief', 'he was a fellow horse thief']
['no']
['yes']
['he was lynched', 'he said he would be lynched', 'by Jeff Jones']
['no']
['that he is a downright horse thief', 'he was a downright horse thief', 'that he was a downright horse thief', 'you are a downright horse thief', 'that you are a down

 45%|████▌     | 45/100 [12:38<14:58, 16.34s/it]

['he was lynched', 'he will be lynched', 'he should be lynched']
['Bonn', 'The Federal City of Bonn']
['in the Rhine-Rhine in the German state of North Rhine-Westphalia', 'on the banks of the Rhine in the German state of North Rhine-Westphalia', 'in the Rhine-Westphalia', 'in the German state of North Rhine-Westphalia', 'North Rhine-Westphalia']
['in the German state of North Rhine-Westphalia', "Germany's state of North Rhine-Westphalia", 'North Rhine-Westphalia', 'Germany']
['no']
['in the southernmost part of the Rhine-Westphalia', 'in the southernmost part of the Rhine-Ruhr region', 'the southernmost part of the Rhine-Westphalia region', 'the southernmost part of the Rhine-Ruhr region', 'south-southeast of Cologne']
['no', 'No']
['over 11 million', '11 million']
['Bonn', 'the Federal City of Bonn']
['yes']
['in the 1st century BC', '1st century BC', 'the 1st century BC']
['a Roman settlement', 'as a Roman settlement', 'Roman settlement']
['no']
['yes']
['Ludwig van Beethoven']
['fro

 46%|████▌     | 46/100 [12:57<15:34, 17.30s/it]

['no', 'No']
['Aquila']
['Masuccio Torri', '"Masuccio Torri."']
['Seven', 'six', 'Five', 'Six']
['Seven', 'Six.', 'Five', 'Six']
['"Armed men, my lords!"', '"We are betrayed!"']
['We are betrayed!"', '"We are betrayed!"', 'We are betrayed!']
['Ferrabraccio.', 'Ferrabraccio']
['His sword', 'his sword', 'his sword.']
['No.']
['He was our only hope.', 'He is our only hope.']
['The only hope of Babbiano.', 'Babbiano.']
['Gian Maria.', 'Gian Maria,', 'The throne of Gian Maria.', 'the throne of Gian Maria.', 'Gian Maria']


 47%|████▋     | 47/100 [13:14<14:58, 16.96s/it]

['Ferrabraccio.']
['the ice', 'ice']
['a bitter breeze', 'a little bitter breeze']
['Dampier']
['get the mainsail', 'to get the mainsail', 'get the mainsail on', 'get the mainsail on to her', 'got the mainsail on to her', 'to get the mainsail on', 'the mainsail', 'to get the mainsail on to her']
['yes', 'Yes']
['that boat', 'the boat', 'a boat']
['white caps', 'a little warmth']
['they were beginning to feel uneasiness', 'a growing uneasiness', 'uneasiness']
['white caps']
['the ice']
['the _Selache_', 'the _Selache', 'Selache', 'the Selache']
['it had become detached', 'it was detached']
['it had become perceptibly narrower', 'it became perceptibly narrower']
['the ice', 'the ice.']
['yes', 'Yes']
['Wyllard']
['he sent a man up into the foremast shrouds', 'heave the man up into the foremast shrouds', 'sent a man up into the foremast shrouds']
['everything clear']
['Wyllard']


 48%|████▊     | 48/100 [13:36<16:09, 18.64s/it]

['he should get the cover off the second boat', 'Get the cover off the second boat', 'to get the cover off the second boat', 'get the cover off the second boat']
['Multimedia', 'Multimedia devices', 'multimedia']
['Video']
['yes', 'Yes']
['Bob Goldstein']
['July 1966', '1966']
['singer and artist Bob Goldstein', 'Bob Goldstein', 'Bobb Goldsteinn', 'a singer', 'singer and artist', 'a singer and artist']
['at Southampton', 'Southampton']
['Lightworks', '"Lightworks at L\'Oursin"', "Lightworks at L'Oursin", "Light Works at L'Oursin", "LightWorks at L'Oursin"]
['a combination of different content forms', 'content that uses a combination of different content forms', 'content that uses a combination of different content forms such as text, audio, images, animations, video and interactive content.', 'a mix of different content forms']
['text, audio, images, animations, video and interactive content.', 'text, audio, images, videos and interactive content', 'text, audio, images, animations, vid

 49%|████▉     | 49/100 [13:58<16:38, 19.58s/it]

['Iris Sawyer', 'David Sawyer']
['Crooked Creek Company', 'the Crooked Creek Company']
['George']
['yes']
['Harry.', 'Harry']
['Yes.']
['Hetertown']
['about a week', 'Almost a week', 'About a week']
['a week', 'about a week', 'A week', 'About a week']
['George', 'Harry']
['Yes.']
['Yes', 'Yes.']
['George', 'George.']
['Yes.']
['relatives', 'some relatives']
['Yes.']
['No.']
['George']
['Harry.', 'Harry']
['Yes.']


 50%|█████     | 50/100 [14:21<17:09, 20.59s/it]

['slowly', 'horse', 'horses', 'a horse', 'slowly toward him']
['a doctor', 'doctor']
['Royal Liverpool', 'Royal Liverpool hospital', 'at the Royal Liverpool hospital', 'the Royal Liverpool hospital']
['Palestine']
['kidney transplants', 'received kidney transplants', 'kidney transplant', 'transplant', 'he donated kidney', 'donated kidney']
['two', 'Two']
['One kidney', 'kidney and liver', 'kidney and kidney', 'kidney', 'kidney and a kidney']
['family']
['Mohammed Duhair, Nadia, 36, and Mohammed Duhair', 'Mohammed Duhair, Mohammed Duhair', 'Mohammed Duhair and Mohammed Duhair', 'Mohammed Duhair, Mohammed Duhair, Mohammed Duhair', 'Mohammed Duhair, 42', 'Mohammed Duhair, 42, and Mohammed Duhair, 36', 'Mohammed Duhair, Mohammed Duhair,', 'Mohammed Duhair, Mohammed Duhair, and Nadia', 'Mohammed Duhair, 42, and Nadia', 'Mohammed Duhair, Nadia, 36', 'Mohammed Duhair, Mohammed Duhair, Nadia, and Mohammed Duhair', 'Mohammed Duhair and Nadia', 'Mohammed Duhair, 42, and his brother']
['six month

 51%|█████     | 51/100 [14:50<18:48, 23.04s/it]

['yes', 'Yes']
['Robinson Crusoe']
['A great storm', 'a terrible storm', 'a great storm']
['yes', 'Yes']
['on an island', 'On an island', 'Rome']
['No']
['He built a house']
['No']
['A wild man']
['"Friday"', 'Friday']
['A servant', 'Served as a servant']
['Robinson Crusoe']
['Yes']
['a boat', 'A boat']
['Use our own hands to work hard', 'Use your own hands', 'use your own hands', 'Use your own hands to work hard']
['Cry']
['He wanted to be a seaman', 'be a seaman', 'To be a seaman']
['travel around the world']
['Rome']
["Robinson's hands were used to work hard", 'Robinson used his hands to work hard', 'Robinson Crusoe used his hands to work hard', 'Robinson used his hands', 'A cave']


 52%|█████▏    | 52/100 [15:11<17:56, 22.42s/it]

['No']
['CHAPTER VI', 'VI', 'VI.']
['Indian club', 'the Indian club']
['Leo']
['a severe injury', 'A severe injury', 'A severe injury.']
['a severe injury', 'it would have inflicted a severe injury']
['a crowd of performers', 'A crowd of performers', 'crowd of performers']
['Leo']
['Snipper']
['he sprang at Leo', 'He sprang at Leo']
['no', 'No']
['yes', 'Yes']
['Two']
['a stinging slap', 'A stinging slap', 'What did Leo do to the boy?', 'What did Leo do?']
['he sprang at Leo', 'He sprang at Leo', 'Killing the boy', 'He was going to kill the boy', 'he was going to kill the boy']
['“I’m going to teach the boy a lesson', '“I’m going to teach the boy a lesson!”', 'I’ll teach you a lesson!”', '“I’ll teach you a lesson!”', '“I’ll teach the boy a lesson!”', 'I’ll teach you a lesson!']
['No']
['He’s going to teach the boy a lesson!”', 'He’re going to teach the boy a lesson', 'He’s going to teach the boy a lesson', 'He’m going to teach the boy a lesson!”', 'He’d be going to teach the boy a less

 53%|█████▎    | 53/100 [15:37<18:36, 23.75s/it]

['the Indian club was caught']
['Zink Media', 'Zink Media, Inc', 'Zink Media, Inc.']
['Portland, Oregon, US', 'Portland, Oregon,', 'Portland, Oregon']
['30 August 2000']
['8 million', 'over 8 million', 'over 1 million']
['over 4.9 million', 'nearly 4.9 million', 'over 1 million']
['nearly 4.9 million', 'nearly 346,000', 'over 1 million']
['yes', 'Yes']
['Kevin Lewandowski']
['community-built sites', 'the success of community-built sites', 'the success of community-built sites such as Slashdot, eBay, and Open Directory Project']
['electronic music']
['yes']
['in 2003', '2003', 'in 2004']
['to support other genres']


 54%|█████▍    | 54/100 [15:54<16:39, 21.74s/it]

['yes']
['Chinese Buddhism', 'Buddhism']
['Mahayana Buddhism', 'Mahayana']
['China']
['Tang dynasty', 'during the Tang dynasty', 'the Tang dynasty']
['Taoism']
['Vietnam']
['strict self-control, meditation-practice, insight into Buddha-nature, and the personal expression of this insight in daily life', 'Self-control, meditation-practice, insight into Buddha-nature', 'self-control, meditation-practice, insight into Buddha-nature, and the personal expression of this insight in daily life', 'Self-control, meditation-practice, insight into Buddha-nature, and the personal expression of this insight in daily life', 'discipline, meditation-practice, insight into Buddha-nature, and the personal expression of this insight in daily life']
['Self-control, insight into Buddha-nature', 'insight into Buddha-nature']
['insight into Buddha-nature']
['Daily life', 'daily life', 'In daily life', 'In everyday life', 'in daily life', 'Through zazen and interaction with an accomplished teacher.', 'Through 

 55%|█████▌    | 55/100 [16:17<16:25, 21.90s/it]

['Chan Buddhism']
['Charles VIII']
['Naples']
["D'Aubigny"]
['July 7']
['yes']
['Montpensier']
['the Viceroy', 'Viceroy']
['the following year', 'in the following year']
['Pozzuoli']
['Gonzalo de Cordoba']
['the House of Aragon']
['the House of Aragon', 'the army of the House of Aragon']
['Giuffredo Borgia']
['habits', 'the habits']
['yes']
['Doña Sancia']
['Prince of Squillace']
['in Calabria', 'Calabria']
['1496', 'in the spring of 1496']


 56%|█████▌    | 56/100 [16:39<16:14, 22.14s/it]

['the allied forces of Venice', 'the allied forces of Venice and the Church']
['Park House', 'at Park House']
['Norfolk, England', 'Norfolk']
['no']
["the royal family's Sandringham estate", 'the Sandringham estate', 'Sandringham estate']
['Johnnie and Frances', 'Johnnie and Frances Spencer']
['yes']
['three', '3']
['Andrew and Edward']
['a mile away', 'about a mile away']
['the butler']
['to swim', 'swim in the pool', 'to swim in the pool']
['play with animals', 'go outdoors, climb trees and play with animals', 'go outdoors, climbing trees and playing with animals']


 57%|█████▋    | 57/100 [16:53<14:05, 19.66s/it]

['yes']
['barn', 'the barn']
['to reach the chicken pen', 'the chicken pen', 'to get the chicken pen']
['to make scrambled eggs for breakfast', 'her father could make scrambled eggs for breakfast', 'her father could make scrambled eggs', 'to make scrambled eggs']
['four', 'five', 'Five']
['four', 'Four']
['quacking sound', 'quacking', 'a quacking sound']
['a nest']
['brown eggs', 'large brown eggs']
['brown']
['no']
['yes', 'no']


 58%|█████▊    | 58/100 [17:02<11:26, 16.35s/it]

['no']
['Donovan', 'Fred']
['yes', 'Yes']
['very severe names']
['stupid', 'for having been so stupid', 'for having been so stupid as to think it possible', 'for being so stupid', 'he was so stupid', 'because he was stupid', 'for having been stupid as a stupid fool', 'he was stupid', 'because he was so stupid']
['his honesty']
['congratulate him upon what', 'congratulate him']
['yes', 'Yes']
['congratulated him', 'congratulated him upon his innocence', 'congratulated him upon what', 'congratulate him']
['Bill']
['the miners']
['yes', 'Yes']
['making a stock company']
['before returning home', 'after returning home']
['No']
['his clients', 'his clients relative']
['Donovan']
['Donovan']
['a fish', "an' a fish"]
['No']


 59%|█████▉    | 59/100 [17:24<12:22, 18.10s/it]

['Yes']
['to the city']
['Yes']
['yes']
['8', '8th', 'eighth']
['Ridge Road Middle School']
['North Carolina', 'in North Carolina']
['yes']
['basketball']
['yes']
['$ 50, 000', '$ 250, 000', '$ 50 fee', '$ 50', '$ 50, 000', '$50']
['no']


 60%|██████    | 60/100 [17:34<10:26, 15.66s/it]

['no']
['"The Proposal"', 'The Proposal']
['romantic comedy']
['classic romantic comedy', 'traditional romantic comedy', 'romantic comedy', 'the charming, traditional romantic comedy', 'the classic romantic comedy', 'a charming, traditional romantic comedy']
['Ryan Reynolds and Sandra Bullock', 'Sandra Bullock and Ryan Reynolds']
['yes']
['No']
['unknown']
['no', 'No']
['executive assistant']
['a younger man', 'young', 'a young man']
['no', 'No']
['yes']
["because it's a romantic comedy", "it's a romantic comedy"]
['marry him', "they'll marry her", 'marry her']
['sexual-harassment lawsuit', 'a sexual-harassment lawsuit', 'sexual-harassment suit', 'sexual harassment']
['Alaska']
['yes']
["the groom-to-be's family", 'a groom-to-be', "a groom-to-be's family", 'a groom-to-be family']
['no', 'No', 'yes']


 61%|██████    | 61/100 [18:00<12:11, 18.77s/it]

['the 1940s', "the 1940s and the '50s", "the 1940s and '50s"]
['Effective Height Above Average Terrain', 'Height Above Average Terrain', 'Height Above Average terrain', 'Height Above average Terrain', 'Height above average terrain', 'Height Above average terrain']
['EHAAT', 'Effective Height Above Average Terrain', 'Effective Height Above average terrain', 'Effective Height Above average', 'HAAT']
['Yes']
['in mountainous regions', 'In mountainous regions']
['FM radio and television', 'Radio and television']
['FM radio and UHF', 'VHF and UHF', 'UHF and UHF']
['No']
['Meter', 'meters']
['It is more important than effective radiated power', 'Because it is more important than effective radiated power', 'Because it is more important than radiated power']


 62%|██████▏   | 62/100 [18:10<10:13, 16.15s/it]

['The maximum distance their station is allowed to cover', 'The maximum distance their antenna class is allowed to cover', 'The maximum distance their station class is allowed to cover']
['Michael Brewer']
['He was burned', 'burned']
['65%', '65 percent', 'Over 65%', '65 percent of his body', 'Over 65 percent of his body', 'Over 65 percent']
['unknown']
['A group of teenagers', 'a group of teenagers']
['No']
['unknown', 'Two years', 'About two years']
['Valerie', 'Valerie Brewer']
['Yes']
['Their son.', 'His younger brother is his older brother.', 'Her brother is his brother.', "Their son is Michael's older brother.", 'Their son is his brother.', 'Their son is their son.']
['CNN']
['Second- and third-degree burns', 'Second- and third-degree burns.']
['Dr. Carl Schulman.', 'Dr. Carl Schulman']
['Dr. Carl Schulman.', 'Carl Schulman.', 'Carl Schulman']
['Jackson Memorial Hospital', 'Jackson Memorial Hospital.', 'Jackson Memorial Hospital Burn Center.']
['About two-thirds', 'Two-thirds', '

 63%|██████▎   | 63/100 [18:32<11:04, 17.97s/it]

['focus on that', 'That I can focus on that.']
['French', 'France']
['Majdi Shakoura']
['about 200 meters', '200 meters']
['Palestinian militants']
['no', 'No']
['No']
['Two']
['Yes']
['Hamas naval building', 'a Hamas naval building']


 64%|██████▍   | 64/100 [18:43<09:26, 15.75s/it]

['Yes']
['Windows 8']
["major changes to the operating system's platform and user interface", "major changes to the operating system's platform"]
["perating system's platform", 'perating system']
['proved its user interface', 'proved user experience', 'prove its user experience', 'proved its user experience on tablets', 'proved its user experience', 'improve its user experience']
['what was?', 'what was it?']
['lay programs and dynamically updated content on a grid of tiles', 'lays programs and dynamically updated content on a grid', 'lays programs and dynamically updated content on a grid of tiles', 'lays programs and updates content on a grid of tiles']
['oad and purchase new software', 'oaded and purchasing new software', 'oaded apps', 'oaded and purchased software', 'oad and purchasing new software', 'oads and purchasing new software', 'oaded and purchased new software', 'oaded software']
['what was one of them?', 'what was one?', 'what is one of them?']
['what is the name of the p

 65%|██████▌   | 65/100 [18:50<07:45, 13.30s/it]

['ng the boot process', 'the boot process']
['Bradley Cooper', 'Clint Eastwood', 'Kyle', 'Taya Kyle']
['yes', 'Yes', 'no']
['Taya Kyle']
['March 16']
['she was shot and killed', 'He was shot', 'he was shot and killed', 'shot and killed', 'He was shot and killed']
['Eddie Ray Routh']
['Most successful sniper', 'most successful sniper', 'the most successful sniper', 'Most successful sniper in United States military history']
['yes']
['Clint Eastwood']
['Clint Eastwood']
['yes']
["Kyle's book", "Kyle's bestselling autobiography", "Kyle's autobiography"]
['she wrote to him', 'she wrote a letter', 'she wrote']
['to bury herself under the covers', 'bury herself', 'bury herself under the covers']
['yes']


 66%|██████▌   | 66/100 [19:07<08:01, 14.17s/it]

['life in prison without the possibility of parole', 'life in prison without the possibility of parole.']
['Ronaldo and Ronaldo', 'Messi and Ronaldo', 'Messi and Cristiano Ronaldo', 'Ronaldo and Messi']
['Ronaldo']
["FIFA Ballon d'Or", "Ballon d'Or"]
['Ronaldo']
['Ronaldo']
['yes', 'Yes']
['thousands of dollars', 'millions of dollars', 'millions']
['Repucom']
['87%', '91%', '92%', '86%']


 67%|██████▋   | 67/100 [19:19<07:24, 13.47s/it]

['87%']
['Rudolph Virchow', 'Virchow']
["Darwin's conclusions lacked empirical foundation", "that Darwin's conclusions lacked empirical foundation", "that Darwin's conclusions lacked foundation", "Darwin's conclusions lacked empirical foundation.", "Darwin's conclusions", "that Darwin's conclusions lacked empirical foundation."]
['1869']
['The Anthropological Society of Madrid', 'the Anthropological Society of Vienna', 'Anthropological Society of Madrid', 'the Anthropological Society of Madrid', 'The Anthropological Society of Vienna']
['the Italian Society of Vienna (1871)', 'the Anthropological Society of Vienna (1871)', 'the Anthropological Society of Vienna', 'the Anthropological Society of Vienna (1870)', 'the Italian Society of Vienna (1870),']
['1871', '1870']
['after that', 'after']
['1902']
['no']
['the study of humans and their societies in the past and the present', 'study of humans and their societies in the past and present', 'a study of humans and their societies in the p

 68%|██████▊   | 68/100 [19:35<07:43, 14.49s/it]

['no']
['3 years old']
['Halloween']
['no']
['a ghost']
['lion', 'a lion']
['no']
['Todd']
['his birthday']
['Mom', 'my mom']
['My mom', 'Mom', 'my Mom', 'my mom', 'my mommy', 'Mommy']
["he'll come home from work", 'he comes home from work', 'he comes home', 'He comes home', 'He comes home from work']
["at his friend Kevin's house", "at Kevin's house", "Kevin's house"]
['my friend', "my friend Kevin's", "Kevin's"]
['because Kevin is my friend', "because he's my friend", 'He is my friend.', "he's my friend", 'he is my friend', "He's my friend", 'He is my friend', 'Kevin is my friend']
['no', 'No']
['Kevin', 'my dad']


 69%|██████▉   | 69/100 [19:51<07:36, 14.73s/it]

['the best Halloween ever', 'the best Halloween ever.']
['Billy']
['to buy some beef', 'to buy beef']
["for his brother's birthday", "his brother's birthday"]
['six', 'Five', 'Six']
['yes']
['brown spots', 'brown']
['eating breakfast', 'eating their breakfast']
['in a big grassy meadow', 'in a grassy meadow', 'a big grassy meadow']
['strange', 'very strange']


 70%|███████   | 70/100 [19:58<06:17, 12.58s/it]

['yes']
['Olivia Pope']
['Jake']
['Fitz', 'a broken Fitz']
['no', 'No']
['Lost his son', 'He lost his son', 'he lost his son', 'lost his son']
['He realized some horrible things about his father', 'realized some horrible things about his father', 'He realized some horrible things about his father.']
['Portia de Rossi', 'Ellen DeGeneres']
['Portia de Rossi']
["Scandal's", 'Scandal']
['Sept. 25 at 9 p.m.', 'Sept. 25']
['9 p.m.,', '9 p.m.', '9 p.m']
['Columbus Short']
['Texas']
["Olivia's", "Olivia's own", 'Olivia Pope']


 71%|███████   | 71/100 [20:12<06:16, 12.97s/it]

['ABC']
['George and Grierson', 'Grierson']
['Grierson']
['Grierson']
["Grant's"]
['frosty', 'snow', 'snow-']
['Grierson']
['George']
['cutting fuel']
['no']
['no']
['trunks', 'the trunks']
['night']
['night']


 72%|███████▏  | 72/100 [20:29<06:34, 14.08s/it]

['the mail-carrier']
['it was more desirable for him', 'more desirable', 'it was more desirable']
['half subdued and hostile to his rule', 'half subdued and hostile']
['his rule']
['England']
['in attacking France']
['by giving them great appointments', 'by giving them great appointments and trusting them fully']
['by calling them to his Parliament', 'by calling them to his Parliament in London']
['run up for her a new constitution', 'run up for her new constitution', 'he ran up for her a new constitution', 'ran up for her a new constitution', 'he hastily ran up for her a new constitution']
['yes']
['John of Brittany']
['as governor', 'a governor', 'governor']
['John of Brittany']
['Scotland']
['the Scottish clergy']
['union with England']
['the Scottish clergy']
['Bruce', 'Robert Bruce']
['Bishop of St Andrews', 'the Bishop of St Andrews']
['made against him', 'they made against him']


 73%|███████▎  | 73/100 [20:52<07:32, 16.76s/it]

['December 10, 1306', 'February 10, 1306', 'March 10, 1306', 'June 1304', 'in June 1304']
['Winter', 'winter']
['socks', 'some socks']
['Olive the owl', 'Olive']
['in the maple tree', 'the old maple tree', 'in the old maple tree', 'on the maple tree', 'maple tree', 'old maple tree', 'on the old maple tree']
['yes']
['her feet were cold', 'her feet were very cold', 'her legs and toes were very cold', 'her legs and toes were cold']
['the ice creek', 'the icy creek']
['Rose']
['six', 'Six']
["she didn't have socks", 'she had no use for socks', 'she had no use for them']
['no']
['her grandmother and mother', 'her grandmother']
['the big hill', 'on the big hill', 'the farmhouse']
['Henrietta', 'Henrietta the human']
['her powerful beak', 'her beak']


 74%|███████▍  | 74/100 [21:12<07:41, 17.75s/it]

['very happy', 'happy']
['Outside', 'outside', 'A little cat', 'a little cat']
['a can of fish', 'a fish', 'fish']
['fish']
['fell sound asleep', 'fell asleep']
['the apartment', 'in his apartment', 'apartment', 'his apartment', 'in the apartment']
['no']
['thousand miles', 'a thousand miles']
['neighbors']
['the cat had been abandoned by his owner', 'that the cat had been abandoned by his owner', 'the cat had been abandoned']
['his owner']
['Willis']
['five years']
['Willis']
['no']
['it was far more heartbreaking', "it was far more heartbreaking than I'd expected", 'it was more heartbreaking']
['yes']
['he was nice', 'nice', 'how nice it was']
['yes', 'no']
['he was lonely', 'lonely']


 75%|███████▌  | 75/100 [21:35<08:01, 19.28s/it]

['no']
['Slovenia']
['the Republic of Slovenia', 'Republic of Slovenia']
['the Republic of Slovenia', 'the Republic of Ljubljana', 'Slovenia', 'Republic of Slovenia']
['RS']
['in southern Central Europe', 'in the southern Central Europe', 'Central Europe', 'southern Central Europe', 'in the Central Europe']
['Eastern and Southern Europe']
['to the west', 'the west']
['Hungary']
['Hungary']
['Croatia', 'to the northeast']
['yes']
['the southeast', 'the south and southeast']
['2.06 million']
['Ljubljana']
['yes']
['yes']
['the United Nations']


 76%|███████▌  | 76/100 [21:57<08:02, 20.11s/it]

['parliamentary', 'a parliamentary republic']
['Aunt Ada']
['Aunt Ada']
['Gillian']
['Brompton']
["Mr. Clement Underwood's church", 'Mrs. Clement Underwood', 'grandmother', 'her grandmother']
['no']
['yes']
['during the latter part of the time']
['no']


 77%|███████▋  | 77/100 [22:09<06:45, 17.64s/it]

['yes']
['Miss Mohun']
['Arnscombe', 'the district of Arnscombe']
['Mysie']
['claim Miss Prescott', 'to claim Miss Prescott', 'to claim Miss Prescott for a game', 'claim Miss Prescott for a game']
['no', 'No']
['heedless']
['Vera and Paula']
['no']
['delighted']
['lawn tennis']
['yes']
['kittens']
['Raki raki', 'Raki', 'raki']
['Raki raki', 'Raki', 'raki']
['Primrose']
['the Goyle']
['Thekla']
['Thekla']
['her aunt', 'the maiden aunt', 'her maiden aunt']


 78%|███████▊  | 78/100 [22:34<07:15, 19.82s/it]

['a lyre', 'a Greek lyre']
['Michael Scott Moore']
['he was freed', 'freed', 'freed him']
['Somali pirates']
['more than two years', 'two years']
['yes', 'Yes']
['Saunders', 'Marlis Saunders']
['his mother', 'her mother']
['She was elated', 'elated', 'Elated']
['He was free', 'to hear he is free', 'he was free', 'she was told he was free', 'to hear he was free']


 79%|███████▉  | 79/100 [22:44<05:58, 17.06s/it]

['just joyful', 'Joyful', "joyful, I can't describe it", 'joyful', "Joyful, I can't describe it", 'Just joyful']
['Jerry']
['train cars', 'train car', 'a train car', 'in a train car']
['all his life']
['Marge']
['36']
['because she wanted to keep him safe', 'she wanted to keep him safe', 'to keep him safe']
['the people who lived in the town nearby', 'the people in the town nearby', 'people who lived in the town', 'people in the town nearby', 'people who lived in the town nearby']
['Qarth']
['100', '100 people']
['the people of Qarth would attack them', 'the people of Qarth would attack them if they knew they were there.', 'the people of Qarth would attack them if they knew they were there']
['corn']
['no', 'No']
['George']
['2 years older than Marge']
["the people in the town didn't like him", "because the people in the town didn't like him", "Because the people in the town didn't like him"]
['yes', 'Yes']
['yes']


 80%|████████  | 80/100 [23:02<05:47, 17.40s/it]

['yes']
['Three', 'Physical, climatic, and biological', 'Physical, climatic, and biological factors', 'physical, climatic, and biological']
['yes', 'Yes', 'Yes.']
['Yes.']
['Environmental and climatic.', 'Environmental and Biological.', 'Physical, climatic and biological.', 'Physical and climatic.', 'Physical, climatic, and biological.', 'Environmental and biological.', 'Physical, climatic, and biological']
['South Africa.', 'southwestern South Africa.']
['No.']
['Southwest', 'southwestern']
['southwestern']
['No.', 'No']
['"precinctive"', 'Precinctive']
['In "in"', '"in"']
['"the people"', '"the people".']
['"precinctive"', 'Precinctive']
['MacCaughey', 'David Sharp']
['Yes']
['Yes', 'Yes.']
['unknown']
['Cosmopolitan distribution.', 'Cosmopolitan distribution']
['The extreme opposite of Endemism', 'The extreme opposite of endemism', 'Endemism']


 81%|████████  | 81/100 [23:25<06:03, 19.13s/it]

['A island', 'an island', 'An island']
['yes', 'Yes']
['a dog', 'dog']
['a dog.', 'a dog', "Bob's wife", "Bob's wife.", "Her mother and Bob's wife", 'Bob']
['yes']
["Frank's wife", 'Frank', 'a dog']
['a fight', 'the long fight', 'a long fight']
['yes']
["Bob's wife"]
['he passed away', 'passed away']


 82%|████████▏ | 82/100 [23:33<04:40, 15.56s/it]

['a bad heart attack', 'bad heart attack']
['in the park', 'the park']
['yes', 'Yes']
['baseball', 'playing baseball']
['the park']
['two baseballs']
['two']
['cola', 'colas', 'yes']
['ten dollars']
['the park']
['Mike']
['yes', 'Yes']
['yes']
['Sally']


 83%|████████▎ | 83/100 [23:43<03:58, 14.06s/it]

['yes']
['linguist', 'a linguist']
['a new language use', 'a new language']
['Stanford University']
['his girlfriend and their three screaming kids', 'his brother, his girlfriend, and their three screaming kids', 'their three screaming kids', 'his girlfriend and three screaming kids', 'his girlfriend, and their three screaming kids']
['unknown', 'We are in New York', 'New York']
['unknown']
['Netspeak']
['Internet or cell phones']
['linguists']
['Increasing spelling and grammatical mistakes', 'increased spelling and grammatical mistakes', 'spelling and grammatical mistakes']
['linguists']
['David Crystal']
['centuries', 'For centuries', 'for centuries']
['Linguist James', 'James']
['by writing']
['teenagers']
['languages', 'text messaging', 'Text messaging and writing']
['teenagers']
['diary writing']


 84%|████████▍ | 84/100 [24:11<04:51, 18.20s/it]

['linguist']
['Grandpa Peter', 'Peter']
['yes']
['His grandparents', 'his grandparents']
['no']
['a cat']
['no']
['no']
['he had understood that if he wanted to sleep in his bed, he would have to get into it before Tubby did', 'Andy had understood that if he wanted to sleep in his bed, he would have to get into it before Tubby did', 'Andy had understood that if he wanted to sleep in his bed, he would have to get into it before Tubby did.']
['no']


 85%|████████▌ | 85/100 [24:22<03:58, 15.91s/it]

['no']
['fatigued', 'fatigue']
['Philip']
['discussing a plan']
['no']
['to write letters', 'to write']
['writing', 'write', 'write her letters', 'write letters', 'writing her letters']
['began discussing a plan', 'he began discussing a plan']
['offering himself as chief of the constabulary force', 'a chief of the constabulary force', 'of offering himself as chief of the constabulary force', 'chief of the constabulary force', 'offering himself as chief of the constabulary force', 'a plan which had occurred to him of offering himself as chief of the constabulary force', 'of offering himself as chief of the constabulary force']
['constabulary force', 'the constabulary force']
['in the county where Redclyffe', 'in the county where Redclyffe was', 'in the county where Redclyffe was situated', 'the county where Redclyffe was situated']
['no']
['Guy']
['no']
['fatigue']
['yes']
['yes']
['a couple of hours', 'for a couple of hours']
['yes']
['yes']


 86%|████████▌ | 86/100 [24:43<04:05, 17.51s/it]

['yes']
['all']
['ow old is she now?', 'how old is she?', 'ow old is she?', 'ow old was she?']
['how old is Hannah?', 'how old was Hannah?']
["what was her cousin's name?", 'what was her cousin called?']
['how old is she?', 'how old was she?', "what was her cousin's name?"]
["the shark's nose", "e shark's nose", "shark's nose", 'her', 'Hannah', 'him']
['from the water to lie on her back', 'her from the water to lie on his back,"she says', 'from the water to lie on his back', 'her from the water to lie on his back', 'gged from the water to lie on his back']
['from the water to lie on his back', 'her from the water to lie on his back', 'dragged from the water to lie on her back', 'gged from the water to lie on his back', 'her from the water to lie on her back']
["r life to Syb's bravery and the fact that great whites,despite their reputation as man-eaters, typically don't target humans,", "r survival to Syb's bravery and the fact that great whites,despite their reputation as man-eaters, 

 87%|████████▋ | 87/100 [24:56<03:29, 16.11s/it]

['award did Hannah win?', 'award did Hannah owe her?', 'award did Hannah owe her survival to?', 'award did Hannah receive?', 'award did Hannah earn?', 'award did Hannah earn for her bravery?']
['five', 'Five']
['Westchester County', 'Westchester']
['Jonas Bronck']
['the first settlement', 'The New Netherland colony', 'the New Netherland colony', 'New Netherland', 'New Netherland colony']
['in the New Netherland colony', 'the New Netherland colony', 'New Netherland', 'New Netherland colony']
['1639', 'in 1639']
['yes']
['the natives were displaced', 'the natives displaced', 'they were displaced', 'the native Lenape']
['1643']
['Five']
['42 square miles']
['2014', 'in 2014']
['1,442,159', '1,434,159', '1,448,159', '1,438,159']
['No']
['fourth', 'fourth highest']
['hip hop', 'Latin music']
['Hip hop', 'hip hop']
['American South', 'from the American South', 'the American South']
['1643']


 88%|████████▊ | 88/100 [25:15<03:22, 16.91s/it]

['the East River', 'the Harlem River', 'Harlem River', 'The Harlem River']
['Caden Rodgers']
['head injury', 'a head injury']
['yes']
['his mother']
['threw him across the room']
['no']
['meth']
['marijuana']
['yes']
['yes']
['first-degree murder']
['Aurora Sentinel', 'the Aurora Sentinel', 'the Aurora Sentinel.']
['two', '2']
['Los Angeles International Airport']
['to see his dying grandson', 'to see his grandson', 'to see his grandson one last time']
['CNN affiliate KABC', 'CNN', 'CNN affiliate']
['no']
['a pilot', 'pilot']
['Southwest Airlines']


 89%|████████▉ | 89/100 [25:40<03:32, 19.36s/it]

['Southwest spokeswoman', 'a Southwest spokeswoman']
['Joey']
['tooth brush', 'to brush his teeth', 'brush his teeth']
['His mother', 'His mom', 'his mother', 'his mom']
['yes', 'Yes']
['He brushed his teeth', 'Joey brushed his teeth', 'brushed his teeth']
['No']
['Yes']
['He was told to brush his teeth', 'he was told to brush his teeth', "he told his mom that he didn't brush his teeth", "He was told that he didn't brush his teeth", 'he was told he did not brush his teeth']
['He pushed him over and started to cry', 'He pushed the boy over', 'He pushed the boy over and started to cry', 'He pushed the boy over and started crying', 'He pushed the boy', 'He pushed him over']
['He pushed the boy over and started to make fun of him', 'pushed the boy over', 'He pushed the boy over', 'He pushed the boy over and started to cry', 'pushed the boy over and started to cry', 'He pushed him over', 'pushed the boy over and started to cry', 'He pushed the boy over and started to cry', 'He pushed him ov

 90%|█████████ | 90/100 [25:54<02:57, 17.71s/it]

['Yes']
['Virginia governor']
['Virginia']
['Terry McAuliffe']
['Republican Ken Cuccinelli', 'Ken Cuccinelli', 'Bob McDonnell']
['Republican']
['Democrat']
['yes', 'Yes']
['No']
['unknown']
['Tuesday evening.', 'Tuesday evening', 'Tuesday']


 91%|█████████ | 91/100 [26:06<02:24, 16.10s/it]

['Richmond.']
['F**k you', '"F**k you."', "'F**k you.'"]
['Chris Carroll']
["at the scene of Tupac Shakur's 1996 drive-by murder", "Tupac Shakur's 1996 drive-by murder", 'the 1996 drive-by murder', '1996']
['Chris Carroll']
['a sergeant', 'retired sergeant', 'Sergeant']
['Las Vegas Metropolitan Police Department', 'the Las Vegas Metropolitan Police Department']
['a drive-by murder', 'drive-by murder']
['September 7th']
['Orlando Anderson', 'Orlando Anderson Anderson']
['yes']
['with 21-year-old Crips gang member Orlando Anderson', 'Crips gang member Orlando Anderson']
['Death Row Records']
['the lobby of the MGM Grand', 'in the lobby of the MGM Grand casino', 'in the lobby of the MGM Grand', 'a MGM Grand casino', 'the lobby of the MGM Grand casino']
['Mike Tyson fight', 'Mike Tyson']
['no']
['a traffic light', 'the traffic light']
['no']
['Cadillac', 'a Cadillac']
['white']


 92%|█████████▏| 92/100 [26:33<02:35, 19.50s/it]

['a white Cadillac pulled up beside Knight and Shakur while they were stopped', 'a white Cadillac pulled up beside Knight and Shakur', 'a white Cadillac pulled up beside Knight and Shakur while they were stopped at a traffic light', 'the white Cadillac pulled up beside Knight and Shakur while they were stopped at a traffic light', 'a white Cadillac pulled up beside Knight and Shakur while they were stopped at a traffic light and one man began shooting out of the back window.', 'a white Cadillac pulled up', 'a white Cadillac']
['Mayfair', 'in Mayfair']
['Paul']
['Arthur']
['once a day', 'every day']
['his regiment', 'from his regiment', 'regiment']
['Paul']
['London', 'in London']
['Mayfair', 'in Mayfair']
['shook hands', 'they shook hands']


 93%|█████████▎| 93/100 [26:46<02:01, 17.40s/it]

['pale cheeks', 'a lot of pale cheeks', 'sunken, pale cheeks,', 'sunken, pale cheeks', 'He was still in his dressing-gown', 'he was in his dressing-gown', 'he was still in his dressing-gown']
['Whiskers has a white spot on her chest', 'Whiskers is black with a white spot on her chest', 'Whiskers is black', 'Whiskers is black with a white spot on her chest.', 'on Saturday, Whiskers turns two years old', 'on Saturday, Whiskers turns two years old.', 'because on Saturday, Whiskers turns two years old.', 'because on Saturday, Whiskers turns two years old']
['a cat', 'a pet cat', 'A pet cat', 'cat', 'A cat']
['like little white mittens', 'a little white mitten', 'little white mittens']
['pet store', 'the pet store']
["to buy Whiskers' birthday presents", "Whiskers' birthday presents"]
['a play mouse and a feather', 'a play mouse', 'a play mouse and a blue feather']
['a blue feather']
['a ball of yarn', 'a red ball of yarn']
['a bowl with a picture of a cat', 'a bowl with a picture of a cat 

 94%|█████████▍| 94/100 [26:55<01:30, 15.09s/it]

['in the sun']
['not using "good judgment,"', 'not using good judgment', 'not using "good judgment"']
['not got out of that car', 'not gotten out of that car', 'not have gotten out of that car', 'not gotten out of the car', 'got out of the car', 'got out of that car', 'gotten out of that car', 'not have been in the car']
['no']
['the car', 'in the car']
["he didn't use good judgment", 'he didn\'t use "good judgment"']
['these people', 'the people']
['George Zimmerman', 'Zimmerman', 'the Zimmerman family', 'the Zimmerman']
['he went above and beyond what he should have done', 'he went above and beyond what he really should have done," she said.', 'he went above and beyond what he really should have done']
['yes']
['Juror B37']
["CNN's Anderson Cooper 360", 'Anderson Cooper 360', 'Anderson Cooper', 'CNN\'s "Anderson Cooper 360"']
['no']
['Monday night']
['yes']
['he was shot', 'he got shot', 'he shot him']
["he didn't use good judgment", 'he didn\'t use "good judgment"']
['no']
["he didn

 95%|█████████▌| 95/100 [27:17<01:25, 17.02s/it]

['in the neighborhoods', 'in the neighborhood']
['reviewing films', 'reviewing']
['Tom Seaton']
['no', 'No']
['Tom Seaton']
['editor of The Front Page', 'first arts editor of The Front Page', 'arts editor', 'first arts editor', 'The first arts editor of The Front Page', 'the first arts editor of The Front Page', 'the first editor of The Front Page']
['written for television', 'he had written for television.', 'he had also written for television', 'he had written for television', 'writing for television']
['no']
['yes']
['yes']


 96%|█████████▌| 96/100 [27:31<01:03, 15.97s/it]

['to make the atmosphere sociable', 'to make sociable atmosphere', 'there would also be guests to make the atmosphere sociable', 'there would also be guests to make sociable', 'there would be a guest to make the atmosphere sociable']
['Safedom']
['European partners or acquisitions']
['Brian Fu']
['200m']
['1bn']
["the world's biggest player", "the world's largest player"]
['Three', 'three']
['chief executive']
['UK', 'the UK', 'in the UK']
['yes']
['meeting potential partners', 'met potential partners and acquisitions', 'meeting potential partners and acquisitions']
['condom', 'condoms']
['Beautiful Girl, Take Me and Green Lemon']
['it is in manufacturing and technology', 'manufacturing', "it's in manufacturing and technology", 'manufacturing and technology', 'in manufacturing and technology']
['condoms', 'half of all condoms are bought by women', 'half of all condoms']
['200m', 'half of all condoms are bought by women', 'half of all condoms']
['branding part', 'branding', 'the brandin

 97%|█████████▋| 97/100 [27:50<00:51, 17.08s/it]

['No']
['Foy']
['the naked back of Martin', 'the back of Martin']
["the shelter of Mother Martha's lair", "Mother Martha's lair"]
['the forced marriage', 'forced marriage']
['in the Red Mill', 'the Red Mill']
['weeks', 'some weeks']
['the sword']
['the Haarlemer Meer', 'Haarlemer Meer', 'in the Haarlemer Meer']
['Lysbeth']
['many days']
['the sword cut', 'the sword cut in his thigh', 'the sword was cut in his thigh', 'sword cut in his thigh']
['yes']
['gangrene']
['yes']
['his own strength and healthy constitution']
['Leyden']


 98%|█████████▊| 98/100 [28:08<00:34, 17.26s/it]

['for him to visit', 'to visit']
['late twentieth-century development', 'Late twentieth-century', 'late twentieth-century', 'late twentieth century development', 'late twentieth century', 'Late twentieth-century development']
['Limited Overs Internationals', 'Limited overs International', 'Limited overs internationals', 'Limited overs cricket']
['Limited Overs Internationals', 'Limited Overs', 'Limited overs internationals', 'Limited Overs International', 'Limited Overs Internationals Internationals']
['Twenty20 International matches', 'Twenty20 Internationals', 'Twenty20 International']
['One Day International', 'A One Day International', 'One Day International matches']
['5 January 1971', '1971']
['5 January 1971', '1971']
['Melbourne', 'Melbourne Cricket Ground']
['no', 'No']
['Australia and England']
['Australia']
['England']
['40 wickets', '5 wickets']
['a red ball.', 'a red ball']
['the rival World Series Cricket competition', 'World Series Cricket', 'World Series Cricket competi

 99%|█████████▉| 99/100 [28:32<00:19, 19.25s/it]

['2001.', '2001']
['the long night', 'a long night', 'long night']
['Billy']
['Byrne']
['he was a mucker', 'a mucker', 'mucker']
['no', 'No']
['Anthony Harding']
['Mallory']
['Mallory']
['Mallory']


100%|██████████| 100/100 [28:42<00:00, 17.23s/it]

['he was sure that the old love would be aroused', "he was sure that the old love would be aroused in the girl's breast", 'he was a prospective son-in-law']
Epoch: 6  with accuracy: 0.7346809854706254


In [21]:
accuracy

0.6910928616550853

## save_small 6
### num_replicas=5, len(dlist)=10 
Epoch: 6  with accuracy: 0.8439716312056738

## save_adversarial 6
### num_replicas=5, len(dlist)=100
Epoch: 6  with accuracy: 0.8313329121920404

### num_replicas=7, len(dlist)=100
Epoch: 6  with accuracy: 0.8458622867972204

### num_replicas=10, len(dlist)=100
Epoch: 6  with accuracy: 0.868603916614024

### num_replicas=20, len(dlist)=100
0.8989260897030954

## Using only most frequent answer
### num_replicas=25, len(dlist)=10
0.7092198581560284

## Using the whole group of most frequent answers
### num_replicas=25, len(dlist)=10
0.7659574468085106
### num_replicas=25, len(dlist)=100
0.7195198989260897

## Using the first two groups of most frequent answers
### num_replicas=25, len(dlist)=100
0.8439671509791535

## Using all groups over a threshold
### num_replicas=25, len(dlist)=100, threshold>0.4
0.6910928616550853

### num_replicas=25, len(dlist)=100, score/best_prediction > 0.80.
0.7346809854706254

## TEST WITH 2 groups and fixed YES/NO answers



# TODO
#### USE DROPOUT ON A BATCH OF IDENTICAL INPUTS!
This way there is no need to run the model 5-6-7 times

In [114]:
checkpoint = torch.load('save_adversarial' + str(1))
model.load_state_dict(checkpoint['model_state_dict'])

<All keys matched successfully>

In [105]:
def subfinder(lst, pattern, offset=0):
    index1 = offset
    index2 = offset
    
    while index1 < len(lst):
        if lst[index1] != pattern[index2]:
            index2 = 0
            index1 += 1
            continue
        index1 += 1
        index2 += 1
        if index2 == len(pattern):
            return index1 - index2
        
    return -1

In [86]:
def generate_answer_with_attention(model, prompt):
    tokens = tokenizer.encode(prompt, return_tensors='pt')
    _length = 50
    tokens_length = tokens.shape[1]
    if tokens_length + _length > 1024:
        return ''
    output = model.generate(
             tokens.cuda(),
             max_length=tokens_length + _length,
             #temperature=0,
             pad_token_id=50256
    )
    output = tokenizer.decode(output[0], skip_special_tokens=True)
    offset = len(prompt)
    start = offset + 1
    end = output.find('\n', start)
    att_output = model(tokens.cuda(), output_attentions=True)
    weights = torch.mean(att_output.attentions[11][0][2][-4:-1], dim=-2)
    return output[start:end].split(':')[-1].strip(), weights, tokens

In [115]:
def generate_multiple_answer_with_attention(model, prompt, num_replicas=10):
    model.train()
    outputs = []
    with torch.no_grad():
        tokens = tokenizer.encode(prompt, return_tensors='pt')
        tokens = tokens.repeat(num_replicas,1)
        _length = 50
        tokens_length = tokens.shape[1]
        if tokens_length + _length > 1024:
            return ''

        
        output = model.generate(
             tokens.cuda(),
             max_length=tokens_length + _length,
             pad_token_id=50256
        )
        for index in range(num_replicas):
            text = tokenizer.decode(output[index, :], skip_special_tokens=True)
            offset = len(prompt)
            start = offset + 1
            end = text.find('\n', start)
            att_output = model(tokens.cuda(), output_attentions=True)
            weights = torch.mean(att_output.attentions[11][0][2][-4:-1], dim=-2)
            outputs.append((text[start:end].split(':')[-1].strip(), weights, tokens))

    return outputs

## TODO IMPROVEMENTS
* Do not just use layer 11
* Average among all "similar" answers (same group)

In [149]:
small_text = get_text_from_data_item(dev_dict['data'][0], 
                                                 max_num_questions=8,
                                                 question_number=0,
                                                 last_question=False)
tokens = tokenizer.encode(small_text, return_tensors='pt')
answer_weights_tokens_list = generate_multiple_answer_with_attention(model, small_text)

In [150]:
words = [tokenizer.convert_tokens_to_string(
         tokenizer.convert_ids_to_tokens([item], skip_special_tokens=True)[0])
         for item in tokens[0]]

In [161]:
answer_weights_tokens_list[7]

('orange',
 tensor([3.6089e-01, 8.1199e-05, 1.0805e-04, 6.5980e-05, 8.9267e-04, 3.7233e-05,
         0.0000e+00, 1.0273e-04, 8.4571e-05, 6.3877e-05, 1.2993e-04, 7.8752e-04,
         1.6993e-04, 7.4322e-06, 8.6734e-04, 4.2481e-03, 3.9622e-03, 5.9246e-04,
         9.9492e-05, 1.0623e-04, 7.9502e-04, 2.6960e-04, 1.0184e-04, 5.8681e-03,
         9.3598e-04, 4.1873e-05, 5.5301e-04, 4.0890e-04, 1.5313e-03, 1.4488e-03,
         1.1528e-02, 1.7011e-03, 8.0181e-04, 5.3033e-03, 4.9567e-02, 1.8595e-03,
         1.9864e-02, 5.4416e-03, 4.1015e-03, 8.2716e-04, 1.9289e-03, 7.8403e-05,
         1.2960e-04, 1.7088e-04, 1.0379e-03, 3.4712e-03, 2.4533e-04, 4.0942e-04,
         8.7080e-05, 8.9755e-04, 1.3617e-03, 2.6039e-04, 8.5770e-05, 1.7864e-04,
         2.7513e-03, 2.0561e-04, 2.9695e-03, 2.6667e-03, 2.8464e-03, 1.2307e-03,
         2.8963e-03, 1.6620e-03, 9.3932e-04, 1.6716e-03, 1.9829e-04, 9.3560e-04,
         2.3852e-04, 2.8511e-04, 1.1969e-03, 1.4051e-05, 1.1094e-04, 6.2961e-04,
         1.8531e-

In [163]:
from colored import colored, fg, bg, attr


answer, weights, tokens = answer_weights_tokens_list[7]

regular_words = [' ', '"', '\'', '.', ',', '."', ':', '?', '!', '".']

start = subfinder(words, ['\n', '\n', 'Story', ':']) + 4
end = subfinder(words, ['\n', '\n', 'Discussion', ':'])
maximum = torch.max(weights[start:end])
minimum = torch.min(weights[start:end])


for word, weight, in zip(words, weights):
    int_color = int(max(0, (1 - (weight - minimum)/(maximum - minimum))) * 20 + 235)
    if int_color < 240:
        int_color = 0
        
    else:
        int_color = 255
        
    if word in regular_words:
        int_color = 255
        
    if int_color > 245:
        color = fg('black') + bg(int_color)
        
    else:
        color = fg('white') + bg(int_color)
        
    res = attr('reset')
    print(color + word + res, end='')
    
    

In the text below two people are discussing a story.

Story:
Once upon a time, in a barn near a farm house, there lived a little white kitten named Cotton. Cotton lived high up in a nice warm place above the barn where all of the farmer's horses slept. But Cotton wasn't alone in her little home above the barn, oh no. She shared her hay bed with her mommy and 5 other sisters. All of her sisters were cute and fluffy, like Cotton. But she was the only white one in the bunch. The rest of her sisters were all orange with beautiful white tiger stripes like Cotton's mommy. Being different made Cotton quite sad. She often wished she looked like the rest of her family. So one day, when Cotton found a can of the old farmer's orange paint, she used it to paint herself like them. When her mommy and sisters found her they started laughing. 

"What are you doing, Cotton?!" 

"I only wanted to be more like you". 

Cotton's mommy rubbed her face on Cotton's and said "Oh Cotton, but your fur is so pret